src/storage/src/sink/kafka.rs

```diff
@@ -90,56 +92,160 @@ where
+        let (purify_stream, purify_errors, purify_token) = purify_connection(
+            format!("kafka-{sink_id}-purify-connection"),
+            input.scope(),
+            &storage_state.connection_context,
             sink_id,
             self.clone(),
+        );
+
+        let (encoded_input, encode_token) = encode_collection(
+            format!("kafka-{sink_id}-{}-encode", self.format.get_format_name()),
+            &input,
+            &purify_stream,
             sink.envelope,
+        );
+
+        let shared_kafka_context = Rc::new(RefCell::new(None));
+
+        let (produce_errors, produce_token) = produce_to_kafka(
+            format!("kafka-{sink_id}"),
+            &encoded_input,
+            &purify_stream,
+            sink_id,
             sink.as_of.clone(),
-            Rc::clone(&shared_frontier),
+            sink.with_snapshot,
             storage_state.sink_metrics.kafka.clone(),
             storage_state
                 .sink_statistics
                 .get(&sink_id)
                 .expect("statistics initialized")
                 .clone(),
             storage_state.connection_context.clone(),
+            Rc::clone(&shared_kafka_context),
+            write_frontier,
         );
 
-        storage_state
-            .sink_write_frontiers
-            .insert(sink_id, shared_frontier);
+        let errors = produce_errors.concat(&purify_errors).flat_map(move |error| {
+            let ssh_status = shared_kafka_context.borrow().as_ref().map(|context| context.tunnel_status());
+
+            let hint: Option<String> = if let ContextCreationError::Other(error) = &*error {
+                match error.downcast_ref::<RDKafkaError>() {
+                    Some(kafka_error) => {
+                        let is_timeout = matches!(kafka_error.code(), RDKafkaErrorCode::OperationTimedOut);
+                        if kafka_error.is_retriable() && is_timeout {
+                            let msg = "If you're running a single Kafka broker, ensure that the \
+                                configs transaction.state.log.replication.factor, \
+                                transaction.state.log.min.isr, and offsets.topic.replication.factor \
+                                are set to 1 on the broker";
+                            Some(msg.to_string())
+                        } else {
+                            None
+                        }
+                    }
+                    None => None,
+                }
+            } else {
+                None
+            };
+
+            let mut statuses = vec![];
 
-        (health, Some(token))
+            if let Some(SshTunnelStatus::Errored(e)) = ssh_status {
+                statuses.push(HealthStatusMessage {
+                    index: 0,
+                    update: HealthStatusUpdate::stalled(e, None),
+                    namespace: StatusNamespace::Ssh,
+                });
+            }
```

Comment:
Not covered by any test according to coverage data (but I am aware that this code was just moved).